decisions/2024-04-15_decimal_pack_size.md

```diff
@@ -0,0 +1,131 @@
+- Calculations will accrue inaccuracy. For individual lines this is usually tolerable but it does accumulate! In tenders we see it in monetary values: summing the "adjust cost" of several thousand lines can result in whole dollars of inaccuracy. Governments/Procurement agencies may have policies of acceptable level of calculation inaccuracy.
+  - It is worth highlighting that number of packs and our monetary fields are all already floating point type with these same problems, so we must already handle all the above issues regardless. **So it's not really a con if we have to do it anyway.** It just makes the surface area of floating point accuracy a smidgen higher.
+- recurring numbers create headaches of who truly has the "full" amount of something. E.g
+
+### Option 2 - Round pack size up to next whole number and adjust number of packs, monetary fields etc.
+
+We convert decimal pack sizes up to the next whole number.
+
+E.g.
+
+1 pack of 1L is repacked to 40 packs of 0.05L in mSupply desktop. The latter on integration in OMS:
+
+1. pack size: 0.05L/0.05 = 1L
+2. number of packs: 40packs\*0.05 = 1pack
+3. cost/sell price: 0.5/0.05 = $10
+
+The pharmacist feedback I have received on this is that it'd be confusing to users if they had a system telling them they had 1 of when they can see there are 40 smaller vessels on the shelf. For accounting, I have not asked a finance stakeholder yet but I presume this is very very bad form in regards of adjusting the number of what was physically ordered/received and the cost of each item ordered.
+
+A potential solution for this is to use pack variants. After converting back to 1L pack size, we use a variant of "50mL bottle" that converts all the properties back in the UI layer. Pack variants don't appear to quite support this - the assumption is the base unit is a small whole number (integer). In this case we'd need a pack variant of 0.05L, which would do the reverse of the calculations above. There is additionally an issue of when and how these pack variants are generated, as they are centrally controlled data.
+
+Going to the option 1 problem statement, we also have to solve this at integration still so no avoided in anyway:
+
+> There is potential for discrepancies to occur, e.g. store A repacks `1 * 1L` to `3 * 0.33L`. There are 0.01 packs of the original line... do we just zero it? If we issue all 3 to another store, they receive 0.99L. Do we round it up? What's the threshold?
+
+_Pros:_
+
+- Open mSupply can keep using integer for pack sizes, which feels good in a programming lens. Floats with their inaccuracy are genuinely not ideal!
+
+_Cons:_
+
+- Unintuitive for users, stock levels in system no longer represents physical reality.
+  - Could be assuaged by making pack variants support decimals by being f64 rather than i32. Is this just shifting the problem though? We'd still need to handle when and if rounding happens in UI and calculations no matter what we do.
```

Comment:
I actually think that rounding is something that we don't need to consider, as the user wouldn't divide things that way ? If it happens it's most likely use error, I think 'decimal' system should support most division without repeating or very large decimal numbers ? (I guess you can be converting from imperial to metric ? ohh man, don't want to thing about this lol)